### PR TITLE
CBG-1317: Pass xattrs into Sync Function

### PR DESCRIPTION
**Pull request by JRascagneres**
- Make function to marshal user xattrs into meta map
- Pass marshalled meta map into Sync Function
- Add tests to ensure data in Sync Function can be used (and doesn't break existing uses)

- [x] http://uberjenkins.sc.couchbase.com:8080/view/All/job/sync-gateway-integration/716
- [x] Based onto https://github.com/couchbase/sync_gateway/pull/4963 so that needs merge and then rebase first

### CODE REVIEW

**Comment by bbrks** [comment]
This is difficult to review ahead of #4963 unless this PR is pointing towards that PR's branch - I would advise doing that to allow reviewing, then switch back to master after rebasing.

**Comment by JRascagneres** [comment]
> This is difficult to review ahead of #4963 unless this PR is pointing towards that PR's branch - I would advise doing that to allow reviewing, then switch back to master after rebasing.

done

File: db/document.go [hunk 1]
```diff
@@ -316,6 +316,22 @@ func (doc *Document) BodyBytes() ([]byte, error) {
 	return doc._rawBody, nil
 }
 
+func (doc *Document) GetMetaMap(userXattrKey string) (map[string]interface{}, error) {
```
Comment (adamcfraser) [inline db/document.go:322]: Can you add a comment describing what this function does?  The name is a bit ambiguous.

File: db/document.go [hunk 2]
```diff
@@ -316,6 +316,22 @@ func (doc *Document) BodyBytes() ([]byte, error) {
...
+	var userXattr interface{}
+	if len(doc.rawUserXattr) > 0 {
+		err := base.JSONUnmarshal(doc.rawUserXattr, &userXattr)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return map[string]interface{}{
```
Comment (adamcfraser) [inline db/document.go:334]: The updated unit tests are passing a nil value for metaMap to handle the case where userXattrKey isn't set, but this looks like it's going to return a map with an empty key and null value.  That doesn't seem desirable.  

I think an early return when userXattrKey is empty also makes sense.

File: rest/api_test.go [hunk 3]
```diff
@@ -5619,3 +5619,46 @@ func TestUptimeStat(t *testing.T) {
...
+	assert.NoError(t, err)
+
+	res = rt.SendAdminRequest("PUT", "/db/doc?rev="+doc.SyncData.CurrentRev, "{}")
+	assertStatus(t, res, http.StatusCreated)
+
+	res = rt.SendAdminRequest("GET", "/db/_raw/doc", "")
+	assertStatus(t, res, http.StatusOK)
+	err = json.Unmarshal(res.BodyBytes(), &doc)
+	require.NoError(t, err)
+	assert.Equal(t, []string{"channel"}, doc.SyncData.Channels.KeySet())
```
Comment (adamcfraser) [inline rest/api_test.go:5663]: Test coverage for the following would be useful:
- updating the doc to remove myxattr
- assigning a channel based on the non-existence of myxattr
- attempting to write a user xattr named "" (not sure if that's supported by kv).  If so, ensure it's not used if UserXattrKey is undefined

**Review by adamcfraser — CHANGES_REQUESTED** [review]

File: rest/api_test.go (hunk 3 above)
Comment (JRascagneres) [reply rest/api_test.go:5663]: @adamcfraser The first two are done as part of the imports PR. I figured it was better to write those tests when I had the ability to properly import them.

File: db/document.go (hunk 2 above)
Comment (JRascagneres) [reply db/document.go:334]: Yeah it does leave it as null. I can early exit. If we want to.

File: rest/api_test.go (hunk 3 above)
Comment (JRascagneres) [reply rest/api_test.go:5663]: For the third one:
- Tested this and KV doesn't allow xattrs to have an empty key so we shouldn't run into this.
- GetWithXattr doesn't obtain any xattrs if the UserXattrKey == ""

File: channels/channelmapper_test.go [hunk 4]
```diff
@@ -46,111 +47,111 @@ func TestOttoValueToStringArray(t *testing.T) {
 // verify that our version of Otto treats JSON parsed arrays like real arrays
 func TestJavaScriptWorks(t *testing.T) {
 	mapper := NewChannelMapper(`function(doc) {channel(doc.x.concat(doc.y));}`)
-	res, err := mapper.MapToChannelsAndAccess(parse(`{"x":["abc"],"y":["xyz"]}`), `{}`, noUser)
+	res, err := mapper.MapToChannelsAndAccess(parse(`{"x":["abc"],"y":["xyz"]}`), `{}`, nil, noUser)
```
Comment (adamcfraser) [inline channels/channelmapper_test.go]: As discussed, pass 'empty meta' to these tests instead of nil.

**Review by adamcfraser — APPROVED** [review]

**Pull request merged**